### DISCUSSION TITLE
Issue with task.executor and HA crashing

### DISCUSSION THREAD

**Discussion by afairclo** [custom-components/pyscript#359] [Q&A] [answered]
Hello,

I've been working on a little script to keep up a Twitch stream on one of my TVs. This requires use of the requests module to interact with the Twitch API. I'm not super proficient in Python, but I think I've got the code right as I'm now locking up my HA installation instead of getting syntax errors. The issue now is when the script is running, HA becomes unresponsive until a core restart is issued or HA just crashes and restarts itself. The logs seem to be cleared on restart, so I'm having difficulty pinning down the issue.

Any advice would be appreciated.

The script:
```
@service
def twitch_picker(action=None, id=None):
    log.info(f"twitch_picker: got action {action} id {id}")
    import json
    import requests
    import random
    
    #entity_id = data.get("entity_id")
    entity_id = "media_player.man_cave_2"
    
    streamers = []
    gameIDs = [
        '497440', # Hell Let Loose
        '493388', # Foxhole
        '941530474', # Mario Kart 8 Deluxe
        '512953', # Elden Ring
        '15467', # Battlefield 2
        '16348', # Battlefield 2142
        '18122', # World of Warcraft
        '508455', # Valheim
        '8091', # Darts
        '130942', # Factorio
        '394568', # Rimworld
        '490959' # Zwift
        ]
    
    # ID list:
    # https://raw.githubusercontent.com/Nerothos/TwithGameList/master/game_info.csv
    
    skipList = [
        'presscorps',
        'bf2tv'
        ]

    # Get stream candidates, top 10 English streamers for each game
    for game in gameIDs:
        requestURL = "https://api.twitch.tv/helix/streams?game_id="+game+"&language=en&first=10"
        headers = {'Authorization':'Bearer redacted','Client-Id':'redacted'}
        response = task.executor(requests.get, requestURL, headers=headers)
        streams = response.json()
        for stream in streams['data']:
            if stream['user_login'] not in skipList:
                streamers.append(stream['user_login'])
    
    
    # print(streamers)
    
    pick = random.choice(streamers)
    
    url = "https://www.twitch.tv/"+pick
    
    hass.services.call("media_extractor", "play_media", {"entity_id": entity_id, "media_content_id": url, "media_content_type": "video"})
```

**Comment by craigbarratt** [comment, marked as answer]
The symptoms sound like something you are calling from pyscript is blocking, which will block the main HA thread.  The obvious candidate is the `requests.get`, but you are correctly calling that via `task.executor`, so that shouldn't be the problem.

Does the stalling of HA only happen after you call the `twitch_picker` service?  Could there be an infinite loop with it being called repeatedly (ie, does a side-effect of calling it cause another trigger to call it)?  The fact that HA actually crashes, rather than just slwoing down, suggests this could be the issue.

The only other call that your function makes is `hass.services.call()`.  That shouldn't block, but I'm not 100% sure.  I'd recommend using the native `pyscript` service call (which actually calls `async_call()`, not `call()`):
```python
service.call("media_extractor", "play_media", entity_id=entity_id, media_content_id=url, media_content_type="video")
```
There are several more intuitive ways of calling a service - see the docs.

If that doesn't help, I'd recommend. bisecting the function - remove code until it no longer blocks so you can figure out which part of the code is causing the problem.  The first lines to remove are the service call to `media_extractor` and the URL fetching.

**Reply by afairclo** [reply]
Bah. It was the service call at the end. I was so close!!

Now to read up on passing the entity_id in through the automation..

Thanks!
